### PR TITLE
[Indexer] add coin indexing

### PR DESCRIPTION
**Pull request by bowenyang007**
### Description
* current_coin_activities tracks gas burn, withdraw, and deposit events (we artificially create the event for gas burn since it's not a registered event)
* coin_balances and current_coin_balances tracks coins in user's 0x1::coin::CoinStore
* coin_infos tracks a coin's initial registration. Note that coin info cannot change (supply, decimals, type) once it's created. 

### Test Plan
Run locally against testnet
`cargo run -p aptos-node --features "indexer" --release -- -f ./fullnode.yaml`

additional config in fullnode.yaml
```
storage:
    enable_indexer: true

indexer:
    enabled: true
    check_chain_id: true
    emit_every: 1000
    postgres_uri: "postgres://postgres@localhost:5432/indexer_v2"
    processor: "coin_processor"
    fetch_tasks: 10
    processor_tasks: 10
```
image

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4579)

### CODE REVIEW

File: crates/indexer/src/models/coin_models/coin_activities.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    Event as APIEvent, Transaction as APITransaction, TransactionInfo as APITransactionInfo,
+    TransactionPayload, UserTransactionRequest, WriteSetChange as APIWriteSetChange,
+};
+use aptos_types::APTOS_COIN_TYPE;
+use bigdecimal::BigDecimal;
+use field_count::FieldCount;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+const BURN_GAS_EVENT: &str = "0x1::aptos_coin::GasBurnEvent";
```
Comment (gregnazario) [inline crates/indexer/src/models/coin_models/coin_activities.rs:24]: There's no event for this.  Are you making up one for your storage?

**Review by gregnazario — APPROVED** [review]

File: crates/indexer/migrations/2022-09-29-073529_add_coin_tables/up.sql [hunk 2]
```diff
@@ -0,0 +1,72 @@
+-- Your SQL goes here
+-- coin infos. Only first transaction matters
+CREATE TABLE coin_infos (
```
Comment (davidiw) [inline crates/indexer/migrations/2022-09-29-073529_add_coin_tables/up.sql:3]: ugh, why not coin_info is plural

File: crates/indexer/src/models/coin_models/coin_activities.rs [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    pub event_account_address: String,
+    pub event_creation_number: i64,
+    pub event_sequence_number: i64,
+    pub owner_address: String,
+    pub coin_type: String,
+    pub amount: BigDecimal,
+    pub activity_type: String,
+    pub is_gas_fee: bool,
+    pub is_transaction_success: bool,
+    pub entry_function_id_str: Option<String>,
```
Comment (davidiw) [inline crates/indexer/src/models/coin_models/coin_activities.rs:55]: maybe script hash as well?

**Review by davidiw — APPROVED** [review]

File: crates/indexer/src/models/coin_models/coin_activities.rs (hunk 1 above)
Comment (bowenyang007) [reply crates/indexer/src/models/coin_models/coin_activities.rs:24]: yes

File: crates/indexer/src/models/coin_models/coin_activities.rs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                        CoinBalance::from_write_resource(write_resource, txn_version).unwrap(),
+                    )
+                } else {
+                    (None, None)
+                };
+            if let Some(coin_info) = maybe_coin_info {
+                coin_infos.push(coin_info);
+            }
+            if let Some((coin_balance, current_coin_balance, event_to_coin_type)) =
+                maybe_coin_balance_data
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/coin_models/coin_activities.rs:140]: Yea.

File: crates/indexer/src/models/coin_models/coin_activities.rs [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        txn_version: i64,
+        event_to_coin_type: &EventToCoinType,
+        entry_function_id_str: &Option<String>,
+    ) -> Self {
+        let amount = match coin_event {
+            CoinEvent::WithdrawCoinEvent(inner) => inner.amount.clone(),
+            CoinEvent::DepositCoinEvent(inner) => inner.amount.clone(),
+        };
+        let event_move_guid = EventGuidResource {
+            addr: event.guid.account_address.to_string(),
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/coin_models/coin_activities.rs:188]: Hmm I've been doing this the whole time. probably too late to change now unless we have a really good reason to (backfill and all)

File: crates/indexer/src/models/coin_models/coin_infos.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        match &CoinResource::from_write_resource(write_resource, txn_version)? {
+            Some(CoinResource::CoinInfoResource(inner)) => {
+                let coin_info_type = &CoinInfoType::from_move_type(
+                    &write_resource.data.typ.generic_type_params[0],
+                    txn_version,
+                )?;
+                let supply = inner.get_supply(coin_supply_lookup).context(format!(
+                    "supply missing in coin info: {:?}, txn {}, coin_supply_lookup {:?}",
+                    inner, txn_version, coin_supply_lookup
+                ))?;
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/coin_models/coin_infos.rs:53]: I hope so? lol.

File: crates/indexer/migrations/2022-09-29-073529_add_coin_tables/up.sql (hunk 2 above)
Comment (bowenyang007) [reply crates/indexer/migrations/2022-09-29-073529_add_coin_tables/up.sql:3]: hehe. Legacy diesel requirement shit. I don't think that diesel 2.0 still requires it but when you have things like "metadatas" this doesn't seem that bad.

File: crates/indexer/src/models/coin_models/coin_activities.rs [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+        // Get coin info, then coin balances. We can leverage coin balances to get the metadata required for events
+        let txn_version = txn_info.version.0 as i64;
+        for wsc in writesets {
+            let (maybe_coin_info, maybe_coin_balance_data) =
+                if let APIWriteSetChange::WriteResource(write_resource) = wsc {
+                    (
+                        CoinInfo::from_write_resource(write_resource, txn_version, &supply_lookup)
+                            .unwrap(),
+                        CoinBalance::from_write_resource(write_resource, txn_version).unwrap(),
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply crates/indexer/src/models/coin_models/coin_activities.rs]: It's a result<nullable>. I just need to make sure taht there's no error.

**Pull request merged**